### PR TITLE
Bug28780 squashed2

### PR DESCRIPTION
**Pull request by mikeperry-tor**

### CODE REVIEW

File: src/core/or/circuitlist.h [hunk 1]
```diff
@@ -92,41 +92,43 @@
...
  *
  *  This type of circuit is *only* created preemptively and never
  *  on-demand. When an HS operation needs to take place (e.g. connect to an
  *  intro point), these circuits are then cannibalized and repurposed to the
  *  actual needed HS purpose. */
-#define CIRCUIT_PURPOSE_HS_VANGUARDS 23
+#define CIRCUIT_PURPOSE_HS_VANGUARDS 24
 
-#define CIRCUIT_PURPOSE_MAX_ 23
+#define CIRCUIT_PURPOSE_MAX_ 25
```
Comment (nmathewson) [inline src/core/or/circuitlist.h]: Shouldn't this be 24 ?

File: src/core/or/circuitpadding.c [hunk 2]
```diff
@@ -161,6 +161,113 @@ circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx)
...
+ * network-driven cell events.
+ *
+ * After CIRCPAD_DELAY_INFINITE microseconds of silence on a circuit, this
+ * function will no longer hold it open (it will return 0 regardless of
+ * what the machines ask for, and thus circuit_expire_old_circuits_clientside()
+ * will close the circuit after roughly 1.25hr of idle time, maximum,
+ * regardless of the padding machine state.
+ */
+int
+circpad_is_using_circuit_for_padding(circuit_t *circ, int reason)
```
Comment (nmathewson) [inline src/core/or/circuitpadding.c]: It's still wrong for a function with a name like this to make changes on a circuit.  It should really just return 1 or 0, and let the caller invoke another function that adjusts the dirty timestamp and the purpose.

Comment (nmathewson) [reply src/core/or/circuitpadding.c]: (If it's going to have the current functionality, it really needs a different name.)

File: src/core/or/circuitpadding.c [hunk 3]
```diff
@@ -161,6 +161,113 @@ circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx)
...
+      continue; // check next machine
+    } else {
+      log_info(LD_CIRC, "Circuit %d is not marked for close because of a "
+               " pending padding machine.", CIRCUIT_IS_ORIGIN(circ) ?
+               TO_ORIGIN_CIRCUIT(circ)->global_identifier : 0);
+
+      /* If the machine has had no network events at all within the
+       * last circpad_delay_t timespan, it's in some deadlock state.
+       * Tell the caller that we don't own it anymore. */
+      if (circ->padding_info[i]->last_cell_time_sec +
```
Comment (nmathewson) [inline src/core/or/circuitpadding.c]: I like that there is some mechanism of deadlock prevention here, but whom do we expect to call this function in this case?

File: src/core/or/circuitpadding.c [hunk 4]
```diff
@@ -161,6 +161,113 @@ circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx)
...
+
+    /* If the machine does not want to control the circuit close itself, then
+     * check the next machine */
+    if (!circ->padding_machine[i]->manage_circ_lifetime) {
+      continue; // check next machine
+    }
+
+    /* If the machine has reached the END state, we can close. Check next
+     * machine. */
+    if (mi->current_state == CIRCPAD_STATE_END) {
```
Comment (asn-d6) [inline src/core/or/circuitpadding.c]: While we are doing fixups here, there is no  reason to have an `else` statement here IIUC. We can just have the `current_state` check here be yet another `continue` check, and continue without nesting.

**Comment by mikeperry-tor** [comment]
Due to merge conflicts, this PR is closed. See #1023. I will still reply to the review comments here, though.

File: src/core/or/circuitlist.h (hunk 1 above)
Comment (mikeperry-tor) [reply src/core/or/circuitlist.h]: Yes. Fixed in d9c864b7294273a5dd41e25fae1bd35571ebf52e (in PR #1023).

File: src/core/or/circuitpadding.c (hunk 2 above)
Comment (mikeperry-tor) [reply src/core/or/circuitpadding.c]: How about circpad_marked_circuit_for_padding()? d9c864b7294273a5dd41e25fae1bd35571ebf52e

File: src/core/or/circuitpadding.c (hunk 3 above)
Comment (mikeperry-tor) [reply src/core/or/circuitpadding.c]: Comment clarified in 98d995a68e054dd0a9d36fcbc368bd6c45e22d07 in PR #1023.

File: src/core/or/circuitpadding.c (hunk 4 above)
Comment (mikeperry-tor) [reply src/core/or/circuitpadding.c]: 17765cf2c9e85d8ac0c6b8585486dd7af51c9779 in PR #1023

**Pull request closed without merging**